File: ibkr_show_frontend/src/api/positions.ts
import { request } from './http'
import type { PositionDetailResponse, PositionListResponse, PositionSummaryResponse } from '@/types/positions'

export interface PositionQuery {
  report_date?: string
  symbol?: string
  asset_class?: string
  sort_by?: string
  sort_order?: 'asc' | 'desc'
  page?: number
  page_size?: number
}

export function fetchPositions(params: PositionQuery): Promise<PositionListResponse> {
  const searchParams = new URLSearchParams()

  Object.entries(params).forEach(([key, value]) => {
    if (value !== undefined && value !== '') {
      searchParams.set(key, String(value))
    }
  })

  const queryString = searchParams.toString()
  return request<PositionListResponse>(`/api/positions${queryString ? `?${queryString}` : ''}`)
}

export function fetchPositionSummary(params: Omit<PositionQuery, 'sort_by' | 'sort_order' | 'page' | 'page_size'>): Promise<PositionSummaryResponse> {
  const searchParams = new URLSearchParams()

  Object.entries(params).forEach(([key, value]) => {
    if (value !== undefined && value !== '') {
      searchParams.set(key, String(value))
    }
  })

  const queryString = searchParams.toString()
  return request<PositionSummaryResponse>(`/api/positions/summary${queryString ? `?${queryString}` : ''}`)
}

export function fetchPositionDetail(params: { symbol: string; asset_class?: string | null }): Promise<PositionDetailResponse> {
  const searchParams = new URLSearchParams()
  searchParams.set('symbol', params.symbol)
  if (params.asset_class) {
    searchParams.set('asset_class', params.asset_class)
  }
  return request<PositionDetailResponse>(`/api/positions/detail?${searchParams.toString()}`)
}

File: ibkr_show_backend/app/api/routes/positions.py
from fastapi import APIRouter, Depends, HTTPException, Query, status

from app.api.deps import get_position_service
from app.clients.es_client import ESClientError
from app.schemas.positions import PositionDetailResponse, PositionListResponse, PositionSummaryResponse
from app.services.position_service import PositionService

router = APIRouter(prefix="/positions", tags=["positions"])


@router.get("", response_model=PositionListResponse)
def list_positions(
    report_date: str | None = Query(default=None),
    symbol: str | None = Query(default=None),
    asset_class: str | None = Query(default=None),
    sort_by: str = Query(default="position_value"),
    sort_order: str = Query(default="desc"),
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=20, ge=1, le=200),
    service: PositionService = Depends(get_position_service),
) -> PositionListResponse:
    try:
        return service.list_positions(
            report_date=report_date,
            symbol=symbol,
            asset_class=asset_class,
            sort_by=sort_by,
            sort_order=sort_order,
            page=page,
            page_size=page_size,
        )
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail=str(exc)) from exc
    except ESClientError as exc:
        raise HTTPException(status_code=status.HTTP_503_SERVICE_UNAVAILABLE, detail=str(exc)) from exc


@router.get("/summary", response_model=PositionSummaryResponse)
def get_positions_summary(
    report_date: str | None = Query(default=None),
    symbol: str | None = Query(default=None),
    asset_class: str | None = Query(default=None),
    service: PositionService = Depends(get_position_service),
) -> PositionSummaryResponse:
    try:
        return service.get_positions_summary(
            report_date=report_date,
            symbol=symbol,
            asset_class=asset_class,
        )
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail=str(exc)) from exc
    except ESClientError as exc:
        raise HTTPException(status_code=status.HTTP_503_SERVICE_UNAVAILABLE, detail=str(exc)) from exc


@router.get("/detail", response_model=PositionDetailResponse)
def get_position_detail(
    symbol: str = Query(),
    asset_class: str | None = Query(default=None),
    service: PositionService = Depends(get_position_service),
) -> PositionDetailResponse:
    try:
        return service.get_position_detail(symbol=symbol, asset_class=asset_class)
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail=str(exc)) from exc
    except ESClientError as exc:
        raise HTTPException(status_code=status.HTTP_503_SERVICE_UNAVAILABLE, detail=str(exc)) from exc
